src/appleseed/renderer/modeling/bsdf/microfacetnormalmappinghelper.h

```diff
@@ -0,0 +1,550 @@
+	{
+		foundation::Vector3f wt = build_tangent(wg, wp);
+
+		float wi_dot_wp = cdot(wi, wp);
+		float wi_dot_wt = cdot(wi, wt);
+		float wp_dot_wg = cdot(wp, wg);
+
+		float ap_wi = wi_dot_wp / wp_dot_wg;
+		float at_wi = (wi_dot_wt * std::sqrt(1.0f - std::pow(wp_dot_wg, 2.0f))) / wp_dot_wg;
+
+		return at_wi / (ap_wi + at_wi);
+	}
+
+	float G1(const foundation::Vector3f& wp, const foundation::Vector3f& w, const foundation::Vector3f& wg, const bool w_wp) const
+	{
+		if (foundation::dot(w, wg) <= 0.0f)
+			return 0.0f;
+
+		float H = 1.0f;
+		foundation::Vector3f wt = build_tangent(wg, wp);
+
+		if (w_wp)
+			H = heaviside(foundation::dot(w, wp));
+		else
+			H = heaviside(foundation::dot(w, wt));
+
+		float w_dot_wg = cdot(w, wg);
+		float wp_dot_wg = cdot(wp, wg);
+		float w_dot_wp = cdot(w, wp);
+		float w_dot_wt = cdot(w, wt);
+
+		float ap_w = w_dot_wp / wp_dot_wg;
+		float at_w = (w_dot_wt * std::sqrt(1.0f - std::pow(wp_dot_wg, 2.0f))) / wp_dot_wg;
+
+		return H * std::min(1.0f, w_dot_wg / (ap_w + at_w));
+	}
+};
+
+
+//
+// MicrofacetNormalMappingHelper class implementation.
+//
+
+template <typename BSDFImpl>
+MicrofacetNormalMappingHelper<BSDFImpl>::MicrofacetNormalMappingHelper(
+    const char*                     name,
+    const ParamArray&               params)
+  : BSDFImpl(name, params)
+{
+}
+
+template <typename BSDFImpl>
+void MicrofacetNormalMappingHelper<BSDFImpl>::sample(
+	SamplingContext&                    sampling_context,
+	const void*                         data,
+	const bool                          adjoint,
+	const bool                          cosine_mult,
+	const LocalGeometry&                local_geometry,
+	const foundation::Dual3f&           outgoing,
+	const int                           modes,
+	BSDFSample&                         sample) const
+{
+	// World space original shading normal.
+	foundation::Vector3f original_shading_normal(local_geometry.m_shading_point->get_original_shading_normal());
+	if (foundation::dot(outgoing.get_value(), original_shading_normal) <= 0.0f)
+		return;
+
+	// World space perturbed shading normal.
+	foundation::Vector3f perturbed_shading_normal(local_geometry.m_shading_point->get_shading_normal());
+
+	// Perturbed normal too similar to original. Tangent vector can not be constructed.
+	if (foundation::dot(original_shading_normal, perturbed_shading_normal) > 1.0f - 1e-5)
+	{
+		BSDFImpl::sample(
+			sampling_context,
+			data,
+			adjoint,
+			false,
+			local_geometry,
+			outgoing,
+			modes,
+			sample);
+		return;
+	}
+
+	// World space tangent.
+    foundation::Vector3f tangent = build_tangent(original_shading_normal, perturbed_shading_normal);
+
+    const float shadow_terminator_freq_mult = local_geometry.m_shading_point->get_object_instance().get_render_data().m_shadow_terminator_freq_mult;
+
+	float final_pdf = 0.0f;
+	if (lambda_p(perturbed_shading_normal, outgoing.get_value(), original_shading_normal) > sampling_context.next2<float>())
```

Comment:
In the `if` and `else` block there is a lot of redundant code, which can be reused.

Reply:
I have added `reflect()` function which is reusing code for reflection. Also, I have added `cosine_multiplier()` function which is reusing code for cosine multiplication.
Do you have some other code reuse in mind?